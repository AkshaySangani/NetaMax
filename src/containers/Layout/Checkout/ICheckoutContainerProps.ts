export interface ICheckoutContainerProps {
  /**
   * Include only steps for login flag
   * @type {boolean}
   */
  onlyLoginSteps: boolean;
}
