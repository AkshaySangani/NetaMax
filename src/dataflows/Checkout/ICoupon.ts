export interface ICoupon {
  /**
   * The coupon's id number.
   * @type {string}
   */
  id: string;

  /**
   * The coupon name.
   * @type {string}
   */
  name: string;

  /**
   * The disccount percentaje number.
   * @type {string}
   */
  discountPercentage: string;

  /**
   * The disccount amount number.
   * @type {string}
   */
  discountAmount: string;
}
