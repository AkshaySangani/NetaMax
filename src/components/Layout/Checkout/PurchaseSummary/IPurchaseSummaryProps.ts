import { DeepMap, FieldError, UseFormRegister } from "react-hook-form";

import { ICouponCodeFormValues } from "dataflows/Checkout/CouponCode/ICouponCodeFormValues";
import { ICoupon } from "dataflows/Checkout/ICoupon";

export interface IPurchaseSummaryProps {
  /**
   * The total net price of the items in the cart without discount.
   * @type {number}
   */
  totalNetPrice: number;

  /**
   * Errors for the form fields.
   * @type {DeepMap<Partial<ICouponCodeFormValues>, FieldError | undefined>}
   */
  errors?: DeepMap<Partial<ICouponCodeFormValues>, FieldError | undefined>;

  /**
   * The form register.
   * @type {UseFormRegister<ICouponCodeFormValues>}
   */
  register?: UseFormRegister<ICouponCodeFormValues>;

  /**
   * The code coupon fetch
   * @type {() => void}
   */
  fetchCode?: () => void;

  /**
   * The loader coupon fetch
   * @type {boolean}
   */
  isCheckingCoupon?: boolean;

  /**
   * The array of coupons
   * @type {ICoupon[]}
   */
  coupon?: ICoupon[];

  /**
   * Indicates the discount amount applied
   * @type {number}
   */
  discountAmount: number;

  /**
   * Indicates the user's wallet amount
   * @type {number}
   */
  walletAmount: number;

  /**
   * Indicates the total amount substracting discount, wallet and coupons.
   * @type {number}
   */
  totalAmount: number;

  /**
   * If coupon input is visible
   * @type {boolean}
   */
  couponInputVisible?: boolean;

  /**
   * If pays with monetas
   * @type {boolean}
   */
  payWithMonetas: boolean;

  /**
   * Open/closes coupon input
   * @type {() => void}
   */
  handleCouponVisible?: () => void;
}
