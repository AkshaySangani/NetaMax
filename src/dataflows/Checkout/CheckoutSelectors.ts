import { CheckoutWizardSteps, OrderStatus } from "constants/checkoutConstants";
import { RootState } from "state/store";

import { ICoupon } from "./ICoupon";

/**
 * Selector to get the checkout panel state
 * @param {RootState} state the root state
 * @returns {boolean} true if it is open, false otherwise
 */
export const selectIsOpen = (state: RootState): boolean => state.checkout.isOpen;

/**
 * Selector to get the current step
 * @param {RootState} state the root state
 * @returns {CheckoutWizardSteps} the current step
 */
export const selectCurrentStep = (state: RootState): CheckoutWizardSteps =>
  state.checkout.currentStep;

/**
 * Selector to get the current order status
 * @param {RootState} state the root state
 * @returns {OrderStatus} the current order status
 */
export const selectOrderStatus = (state: RootState): OrderStatus => state.checkout.orderStatus;

/**
 * Selector to get the placing order status
 * @param {RootState} state the root state
 * @returns {boolean} the current placing order status
 */
export const selectPlacingOrderStatus = (state: RootState): boolean =>
  state.checkout.isPlacingOrder;

/**
 * Selector to get the countdown timer
 * @param {RootState} state the root state
 * @returns {boolean} the countdown timer
 */
export const selectCountDown = (state: RootState): number => state.checkout.countDown;

/**
 * Selector to get the placing order status
 * @param {RootState} state the root state
 * @returns {boolean} the current placing order status
 */
export const selectIsCheckingCoupon = (state: RootState): boolean =>
  state.checkout.isCheckingCoupon;

/**
 * Selector to get the countdown timer
 * @param {RootState} state the root state
 * @returns {string} the countdown timer
 */
export const selectCoupon = (state: RootState): ICoupon[] | undefined => state.checkout.coupon;

/**
 * The selector to get the error message
 * @param {RootState} state the root state
 * @returns {ReactElement} the error message
 */
export const selectPlaceOrderErrorMessage = (state: RootState): string | undefined =>
  state.checkout.placeOrderErrorMessage;

/**
 * The selector to get the coupon error message
 * @param {RootState} state the root state
 * @returns {ReactElement} the coupon error message
 */
export const selectCouponErrorMessage = (state: RootState): string | undefined =>
  state.checkout.couponErrorMessage;

/**
 * Selector to get if the user pays with monetas
 * @param {RootState} state the root state
 * @returns {boolean} the pay with monetas flag
 */
export const selectPayWithMonetas = (state: RootState): boolean => state.checkout.payWithMonetas;

/** The selector for the flag to indicate only login steps
 * @param {RootState} state the root state
 * @returns {boolean} the flag to indicate only login steps
 */
export const selectOnlyLoginSteps = (state: RootState): boolean => state.checkout.onlyLoginSteps;
