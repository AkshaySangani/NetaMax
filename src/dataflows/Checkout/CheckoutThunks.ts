import axios, { AxiosError } from "axios";
import { PLACE_ORDER_URL_V2 } from "constants/checkoutConstants";
import { getOrderChallengeStatus } from "dataflows/GanaDinero/GanaDineroThunks";
import { get, post } from "utils/http";

import { createAsyncThunk } from "@reduxjs/toolkit";

import { ICheckoutOrder } from "./ICheckoutOrder";
import { ICoupon } from "./ICoupon";

export const postPlaceOrder = createAsyncThunk(
  "checkout/placeOrder",
  async (order: ICheckoutOrder, { dispatch }) => {
    try {
      //TODO: se cambio PLACE_ORDER_URL por la ruta PLACE_ORDER_URL_V2);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const response = await post(PLACE_ORDER_URL_V2, order as any);
      const orderId = (await response.data) as { id: number };

      dispatch(getOrderChallengeStatus(String(orderId.id)));
    } catch (error) {
      let interalErrorMessage = "No pudimos procesar tu orden. Intenta de nuevo";
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError;
        if (error?.response) {
          interalErrorMessage =
            axiosError.response?.data.message || "No pudimos procesar tu orden. Intenta de nuevo";
        }
      }
      throw Error(interalErrorMessage);
    }
  }
);

export const getCouponCode = createAsyncThunk("checkout/couponCode", async (couponCode: string) => {
  try {
    const response = await get(`discount?code=${couponCode}`);
    return (await response.data) as ICoupon[];
  } catch (error) {
    let interalErrorMessage = "Cup칩n inv치lido";
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      if (error?.response) {
        interalErrorMessage = axiosError.response?.data.error || "Cup칩n inv치lido";
      }
    }
    throw Error(interalErrorMessage);
  }
});
