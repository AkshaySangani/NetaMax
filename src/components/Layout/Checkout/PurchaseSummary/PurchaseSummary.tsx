import { ReactElement } from "react";

import { CURRENCY_NAME } from "constants/localeFormatConstants";
import { formatMoney } from "utils/currencyUtils";

import { Box, Flex, Img, Text } from "@chakra-ui/react";

import { CouponInput } from "../CouponInput/CouponInput";
import { IPurchaseSummaryProps } from "./IPurchaseSummaryProps";

/**
 * The PurchaseSummary component.
 * @param { IOrderReviewSectionProps } props the component props.
 * @returns { ReactElement } The OrderReviewSection component.
 */
export const PurchaseSummary = (props: IPurchaseSummaryProps): ReactElement => {
  const {
    couponInputVisible,
    discountAmount,
    errors,
    fetchCode,
    handleCouponVisible,
    isCheckingCoupon,
    register,
    totalAmount,
    totalNetPrice,
    walletAmount,
    coupon,
    payWithMonetas,
  } = props;
  const actualCoupon = coupon && coupon[0];
  const fromReview = handleCouponVisible && errors && register && fetchCode;

  return (
    <Box marginTop="14px">
      <Flex width="100%" justifyContent="space-between">
        <Text>Suma parcial</Text>
        <Text fontSize="16px" fontWeight="800">
          {formatMoney(totalNetPrice, CURRENCY_NAME)}{" "}
        </Text>
      </Flex>
      <Flex width="100%" justifyContent="space-between">
        <Text>Envío</Text>
        <Text fontSize="16px" fontWeight="800">
          Gratis
        </Text>
      </Flex>
      {actualCoupon && (
        <Flex width="100%" justifyContent="space-between">
          <Flex>
            <Text>Descuento cupón</Text>
          </Flex>
          <Text fontSize="16px" fontWeight="800" color="#F04D4B">
            {discountAmount > 0 && <Text>-{formatMoney(discountAmount, CURRENCY_NAME)}</Text>}
          </Text>
        </Flex>
      )}
      {walletAmount > 0 && payWithMonetas && (
        <Flex width="100%" justifyContent="space-between">
          <Flex>
            <Text>Monetas</Text>
            <Img marginLeft="8px" src="/assets/images/icons/dolar.png" />
          </Flex>
          <Text fontSize="16px" fontWeight="800" color="#F04D4B">
            - {formatMoney(walletAmount, CURRENCY_NAME)}
          </Text>
        </Flex>
      )}
      {fromReview && (
        <CouponInput
          coupon={coupon}
          couponInputVisible={couponInputVisible || false}
          handleCouponVisible={handleCouponVisible}
          errors={errors}
          register={register}
          isCheckingCoupon={isCheckingCoupon || false}
          fetchCode={fetchCode}
        />
      )}
      <Box margin="10px 0" w="100%">
        <hr />
      </Box>
      <Flex width="100%" justifyContent="space-between">
        <Text fontWeight="800" fontSize="16px">
          Total a pagar
        </Text>
        <Text fontSize="16px" fontWeight="800">
          {formatMoney(totalAmount, CURRENCY_NAME)}
        </Text>
      </Flex>
    </Box>
  );
};
