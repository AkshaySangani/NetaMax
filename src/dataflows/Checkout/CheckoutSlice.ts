import {
  CheckoutWizardSteps,
  COUNTDOWN_TIMER_SECONDS,
  OrderStatus,
} from "constants/checkoutConstants";

import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { getCouponCode, postPlaceOrder } from "./CheckoutThunks";
import { ICheckoutState } from "./ICheckoutState";

const initialState: ICheckoutState = {
  isOpen: false,
  onlyLoginSteps: false,
  currentStep: CheckoutWizardSteps.BASKET,
  isPlacingOrder: false,
  orderStatus: OrderStatus.DRAFT,
  countDown: COUNTDOWN_TIMER_SECONDS,
  isCheckingCoupon: false,
  coupon: undefined,
  payWithMonetas: true,
};

const categorySlice = createSlice({
  name: "checkout",
  initialState,
  reducers: {
    removeCouponCode: (state) => {
      state.coupon = undefined;
    },
    setCurrentStep: (state, action: PayloadAction<CheckoutWizardSteps>) => {
      state.currentStep = action.payload;
    },
    setPayWithMonetas: (state, action: PayloadAction<boolean>) => {
      state.payWithMonetas = action.payload;
    },
    moveToNextStep: (state) => {
      const nextStep = state.currentStep + 1;
      state.currentStep = nextStep;
    },
    moveToPreviousStep: (state) => {
      const previousStep = state.currentStep - 1;
      state.currentStep = previousStep;
    },
    onOpen: (state, action: PayloadAction<{ onlyLoginSteps: boolean }>) => {
      state.isOpen = true;
      state.onlyLoginSteps = action.payload.onlyLoginSteps;
    },
    onClose: (state) => {
      state.isOpen = false;
      if (state.orderStatus === OrderStatus.SUBMITTED) state.orderStatus = OrderStatus.DRAFT;
    },
    setCountDown: (state, action: PayloadAction<number>) => {
      state.countDown = action.payload;
    },
    resetCountDown: (state) => {
      state.countDown = COUNTDOWN_TIMER_SECONDS;
    },
    clearPlaceOrderErrorMessage: (state) => {
      state.placeOrderErrorMessage = undefined;
    },
  },
  extraReducers: {
    [postPlaceOrder.pending.type]: (state) => {
      state.isPlacingOrder = true;
    },
    [postPlaceOrder.fulfilled.type]: (state) => {
      state.isPlacingOrder = false;
      state.coupon = undefined;
      state.placeOrderErrorMessage = undefined;
      state.orderStatus = OrderStatus.SUBMITTED;
    },
    [postPlaceOrder.rejected.type]: (state, payload) => {
      state.isPlacingOrder = false;
      state.orderStatus = OrderStatus.REJECTED;
      state.placeOrderErrorMessage = payload.error.message;
    },
    [getCouponCode.pending.type]: (state) => {
      state.isCheckingCoupon = true;
      state.couponErrorMessage = undefined;
    },
    [getCouponCode.fulfilled.type]: (state, action) => {
      state.isCheckingCoupon = false;
      state.coupon = action.payload;
    },
    [getCouponCode.rejected.type]: (state, payload) => {
      state.isCheckingCoupon = false;
      state.couponErrorMessage = payload.error.message;
    },
  },
});

/**
 * Actions
 */
export const {
  setCurrentStep,
  moveToNextStep,
  moveToPreviousStep,
  onClose,
  onOpen,
  setCountDown,
  resetCountDown,
  clearPlaceOrderErrorMessage,
  removeCouponCode,
  setPayWithMonetas,
} = categorySlice.actions;

/**
 * Reducers
 */

export const checkoutReducer = categorySlice.reducer;
