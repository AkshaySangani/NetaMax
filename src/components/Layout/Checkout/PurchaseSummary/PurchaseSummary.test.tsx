import { cleanup, render, screen } from "@testing-library/react";
import { PurchaseSummary } from "./PurchaseSummary";
import { IPurchaseSummaryProps } from "./IPurchaseSummaryProps";
import { ICoupon } from "../../../../dataflows/Checkout/ICoupon";
import { FieldError } from "react-hook-form";

describe("PurchaseSummary component test", () => {
  let mockSuccess: IPurchaseSummaryProps;
  let mockReview: IPurchaseSummaryProps;
  const couponCode: FieldError = { type: "a" };
  const mockedErrors = {
    couponCode,
  };
  const mockedCoupon: ICoupon = {
    id: "1",
    name: "My Coupon",
    discountAmount: "10",
    discountPercentage: "0",
  };
  beforeEach(() => {
    mockSuccess = {
      discountAmount: 10,
      totalAmount: 30,
      totalNetPrice: 60,
      walletAmount: 20,
      coupon: [mockedCoupon],
      payWithMonetas: true,
    };
    mockReview = {
      ...mockSuccess,
      handleCouponVisible: jest.fn(),
      errors: mockedErrors,
      register: jest.fn(),
      fetchCode: jest.fn(),
      couponInputVisible: true,
      isCheckingCoupon: false,
    };
  });

  afterEach(() => {
    cleanup();
  });

  test("PurchaseSummary should be rendered correctly", () => {
    const { container } = render(<PurchaseSummary {...mockSuccess} />);
    expect(container).toMatchSnapshot();
  });

  test("PurchaseTotalAmount should be in the document for success scenario", () => {
    render(<PurchaseSummary {...mockSuccess} />);
    expect(screen.getByText(`$${mockSuccess.totalAmount}.00`)).toBeInTheDocument();
  });

  test("PurchaseTotalAmount should be in the document for review scenario", () => {
    render(<PurchaseSummary {...mockReview} />);
    expect(screen.getByText(`$${mockReview.totalAmount}.00`)).toBeInTheDocument();
  });

  test("PurchaseTotalAmount should be in the document for review scenario", () => {
    render(<PurchaseSummary {...mockSuccess} />);
    const input = screen.queryByPlaceholderText("Escribe aqui tu cupón");
    expect(input).toBeNull();
  });

  test("PurchaseTotalAmount should be in the document for success scenario", () => {
    render(<PurchaseSummary {...mockSuccess} />);
    expect(screen.getByText(`- $${mockSuccess.walletAmount}.00`)).toBeInTheDocument();
  });
});
