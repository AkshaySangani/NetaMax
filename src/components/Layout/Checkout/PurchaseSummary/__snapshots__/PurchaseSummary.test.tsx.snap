// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PurchaseSummary component test PurchaseSummary should be rendered correctly 1`] = `
<div>
  <div
    class="css-wmz062"
  >
    <div
      class="css-z7mtfw"
    >
      <p
        class="chakra-text css-0"
      >
        Suma parcial
      </p>
      <p
        class="chakra-text css-1nhysjy"
      >
        $60.00
         
      </p>
    </div>
    <div
      class="css-z7mtfw"
    >
      <p
        class="chakra-text css-0"
      >
        Envío
      </p>
      <p
        class="chakra-text css-1nhysjy"
      >
        Gratis
      </p>
    </div>
    <div
      class="css-z7mtfw"
    >
      <div
        class="css-k008qs"
      >
        <p
          class="chakra-text css-0"
        >
          Descuento cupón
        </p>
      </div>
      <p
        class="chakra-text css-fa9nx4"
      >
        <p
          class="chakra-text css-0"
        >
          -
          $10.00
        </p>
      </p>
    </div>
    <div
      class="css-z7mtfw"
    >
      <div
        class="css-k008qs"
      >
        <p
          class="chakra-text css-0"
        >
          Monetas
        </p>
        <img
          class="chakra-image css-1isemmb"
          src="/assets/images/icons/dolar.png"
        />
      </div>
      <p
        class="chakra-text css-fa9nx4"
      >
        - 
        $20.00
      </p>
    </div>
    <div
      class="css-eofv09"
    >
      <hr />
    </div>
    <div
      class="css-z7mtfw"
    >
      <p
        class="chakra-text css-ha3xn3"
      >
        Total a pagar
      </p>
      <p
        class="chakra-text css-1nhysjy"
      >
        $30.00
      </p>
    </div>
  </div>
</div>
`;
